script.py

```diff
@@ -0,0 +1,125 @@
+import requests
+import csv
+import json
+import os
+import sys
+from urllib.parse import urlparse
+
+
+def extract_fields(spec, data):
+    """
+    Extract values of fields from data according to a specification.
+
+    Args:
+        spec:   The specification of fields to extract the values from.
+        data:   The data to extract the field values from.
+
+    Returns:
+        An array of tuples, where each tuple contains:
+            * A tuple containing dictionary keys leading to the field
+            * The field value
+    """
+    tuples = []
+
+    if isinstance(spec, dict):
+        for k, v in spec.items():
+            if k in data:
+                for t in extract_fields(v, data[k]):
+                    tuples.append(((k,) + t[0], t[1]))
+    elif isinstance(spec, list):
+        assert len(spec) == 1
+        for idx, d in enumerate(data):
+            tuples += extract_fields(spec[0], d)
+    elif spec is True:
+        return [(tuple(), data)]
+
+    return tuples
+
+
+def get_url_info(url):
+    try:
+        if url.endswith('/'):
+            status_code = 0
+            content_type = ''
+            size = ''
+            extension = ''
+            url_length = sys.getsizeof(url.encode('utf-8'))
+        else:
+            response = requests.head(url)
+            status_code = response.status_code
+            content_type = ''
+            size = ''
+            if status_code == 200:
+                content_type = response.headers.get('Content-Type')
+                size = response.headers.get('Content-Length')
+                extension = os.path.splitext(urlparse(url).path)[1]
+            else:
+                extension = os.path.splitext(urlparse(url).path)[1]
+            url_length = len(url.encode('utf-8'))
+
+        return (status_code, content_type, size, extension, url_length)
+    except requests.exceptions.RequestException:
+        return (0, '', '', '', sys.getsizeof(url.encode('utf-8')))
+
+
```

Comment:
```suggestion
def get_url_info(url):
    status_code = 0
    content_type = ''
    size = ''
    extension = os.path.splitext(urlparse(url).path)[1]
    url_length = sys.getsizeof(url.encode('utf-8'))

    # Some CI systems send links to index pages, instead of actual files,
    # which is against the schema, and we don't really want to cache them.
    # That's why we ignore URLs ending with "/"
    if url.endswith('/'):
        return (status_code, content_type, size, extension, url_length)

    try:
        response = requests.head(url)
    except requests.exceptions.RequestException as error:
        print("Error while trying to get URL info:", url, repr(error))        
    else:
        status_code = response.status_code
    
        if status_code == 200:
            content_type = response.headers.get('Content-Type')
            size = response.headers.get('Content-Length')

    return (status_code, content_type, size, extension, url_length)
```

- Adding a comment explaining why we skip "/"
- Wrapping only the request with try/except
- Extracting lines that repeat in both sides of if/else
- Returning early to avoid long ifs (This is 100% my personal preference, there are people that would prefer always having a single return)

Reply:
@octonawish-akcodes This suggestion includes mostly all the changes I would do to your code to improve its readability (in my opinion)

Reply:
Okay, Now I got it. Thank you :)